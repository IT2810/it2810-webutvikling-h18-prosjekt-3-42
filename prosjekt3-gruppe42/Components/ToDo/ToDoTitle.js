import React from "react";
import { Surface, Button } from "react-native-paper";
import {
  StyleSheet,
  Text,
  View,
  TouchableOpacity,
  DatePickerAndroid,
  AsyncStorage
} from "react-native";

import { MaterialIcons } from "@expo/vector-icons";
import {
  todoBackgroundColor,
  todoEditColor,
  todoDeleteColor,
  todoTitleColor,
  todoDescriptionColor
} from "../../assets/styles";

export default class ToDo extends React.Component {
  constructor(props) {
    super(props);
    this.locale = "nb-no";
    this.months = [
      "januar",
      "februar",
      "mars",
      "april",
      "mai",
      "juni",
      "juli",
      "august",
      "september",
      "oktober",
      "november",
      "desember"
    ];
    this.prioColors = ["#8FC93A", "#E4CC37", "#CC2936"];
  }

  render() {
    var date = this.props.data.date.year != "" ? new Date(this.props.data.date.year, this.props.data.date.month, this.props.data.date.day) : new Date();
    var year = date.getFullYear();
    var month = date.getMonth();
    var day = date.getDate();
    var completed = this.props.data.completed ? (<MaterialIcons name="check" size={20} color={"green"} />) : null
    return (
        <Text
          style={[
            styles.title,
            { borderBottomColor: this.prioColors[this.props.data.priority] }
          ]}>
          {" " +
          this.props.data.title} - Deadline:{" "}
          {day + ". " + this.months[month] + " (" + year + ")"} {completed}{" "}
        </Text>
    );
  }
}

const styles = StyleSheet.create({
  title: {
    borderBottomWidth: 2,
    // borderBottomColor: '#e0e0e0',
    fontSize: 20,
    color: todoTitleColor,
    // padding: 5,
    fontWeight: "bold"
  }
});
