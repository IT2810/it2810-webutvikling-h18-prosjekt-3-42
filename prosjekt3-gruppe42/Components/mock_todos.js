export const todos =  [
    {
      "completed": false,
      "date":{
        "day": 16,
        "month": 9,
        "year": 2018,
      },
      "description": "Dr√∏m",
      "helperText": "Please select at valid location",
      "key": "1539641319117",
      "location":  [
         {
          "accuracy": 0,
          "altitude": 0,
          "latitude": 63.418969700000005,
          "longitude": 10.4309092,
        },
      ],
      "position":  [
         {
          "street": "no position found",
        },
      ],
      "priority": 0,
      "searchbar": "ü¶Ü",
      "searched": false,
      "title": "Sove",
    },
     {
      "completed": false,
      "date":  {
        "day": 16,
        "month": 9,
        "year": 2018,
      },
      "description": "Fiskeburgere",
      "helperText": "You must search for a location before saving (Push the icon)",
      "key": "1539641405341",
      "location":  [
         {
          "accuracy": 0,
          "altitude": 0,
          "latitude": 63.418969700000005,
          "longitude": 10.4309092,
        },
      ],
      "position":  [
         {
          "street": "no position found",
        },
      ],
      "priority": 0,
      "searchbar": "Bunnpris tyholt ",
      "searched": true,
      "title": "Bunnpris",
    },
     {
      "completed": false,
      "date":  {
        "day": 16,
        "month": 9,
        "year": 2018,
      },
      "description": "√Ü",
      "helperText": "Please select at valid location",
      "key": "1539641090267",
      "location":  [
         {
          "accuracy": 0,
          "altitude": 0,
          "latitude": 59.913868799999996,
          "longitude": 10.7522454,
        },
      ],
      "position":  [
         {
          "street": "no position found",
        },
      ],
      "priority": 0,
      "searchbar": "Oslo ",
      "searched": true,
      "title": "Web",
    },
     {
      "completed": false,
      "date":  {
        "day": 17,
        "month": 9,
        "year": 2018,
      },
      "description": "Kj√∏p dopapir",
      "helperText": "You must search for a location before saving (Push the icon)",
      "key": "1539641368526",
      "location":  [
         {
          "accuracy": 0,
          "altitude": 0,
          "latitude": 63.409942699999995,
          "longitude": 10.4389451,
        },
      ],
      "position":  [
         {
          "street": "no position found",
        },
      ],
      "priority": 1,
      "searchbar": "Rema 1000 moholt",
      "searched": true,
      "title": "Rema",
    },
     {
      "completed": true,
      "date":  {
        "day": 16,
        "month": 9,
        "year": 2018,
      },
      "description": "",
      "helperText": "Please select at valid location",
      "key": "1539641937340",
      "location":  [
         {
          "accuracy": 0,
          "altitude": 0,
          "latitude": 63.4402936,
          "longitude": 10.4005057,
        },
      ],
      "position":  [
         {
          "street": "no position found",
        },
      ],
      "priority": 0,
      "searchbar": "Pirbadet Trondheim",
      "searched": true,
      "title": "Bade",
    },
  ]
