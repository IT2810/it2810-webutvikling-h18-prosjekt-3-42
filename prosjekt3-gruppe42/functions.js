// Math stuff
// https://www.movable-type.co.uk/scripts/latlong.html
export const haversine = function (currentLocation, location) {
  const R = 6371e3
  const theta_1 = toRadians(currentLocation.latitude)
  const theta_2 = toRadians(location.latitude)
  const delta_theta = toRadians(location.latitude - currentLocation.latitude)
  const delta_lambda = toRadians(location.longitude - currentLocation.longitude)

  const a = Math.sin(delta_theta / 2) * Math.sin(delta_theta / 2) + Math.cos(theta_1) * Math.cos(theta_2) * Math.sin(delta_lambda / 2) * Math.sin(delta_lambda / 2)

  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a))

  const d = R * c

  return d
}

export const toRadians = function (number) {
  return number * (Math.PI / 180)
}


export function sortKey(x, y) {
  return parseInt(x.key) - parseInt(y.key);
}

export function sortTitle(x, y) {
  return x.title === y.title ? 0 : x.title > y.title ? 1 : -1;
}

export function sortPriority(x, y) {
  return y.priority - x.priority;
}

export function sortDate(x, y) {
  return (
    new Date(x.date.year, x.date.month, x.date.day) -
    new Date(y.date.year, y.date.month, y.date.day)
  );
}
