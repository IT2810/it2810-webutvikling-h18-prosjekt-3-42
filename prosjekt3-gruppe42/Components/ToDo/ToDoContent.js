import React from "react";
import { Surface, Button } from "react-native-paper";
import {
  StyleSheet,
  Text,
  View,
  TouchableOpacity,
  DatePickerAndroid,
  AsyncStorage
} from "react-native";

import { MaterialIcons } from "@expo/vector-icons";
import {
  todoBackgroundColor,
  todoEditColor,
  todoDeleteColor,
  todoTitleColor,
  todoDescriptionColor
} from "../../assets/styles";

import { haversine, toRadians } from "../../functions";

export default class ToDoContent extends React.Component {
  constructor(props) {
    super(props);
    this.locale = "nb-no";
    this.months = [
      "januar",
      "februar",
      "mars",
      "april",
      "mai",
      "juni",
      "juli",
      "august",
      "september",
      "oktober",
      "november",
      "desember"
    ];
    this.prioColors = ["#8FC93A", "#E4CC37", "#CC2936"];
  }

  render() {
    return (
      <View>
        <Text>
          {// Just in case the location data is undefined
          this.props.data.location[0]
            ? "Distance: " +
              (haversine(
                this.props.currentLocation,
                this.props.data.location[0]
              ) > 5000
                ? Math.round(
                    haversine(
                      this.props.currentLocation,
                      this.props.data.location[0]
                    ) / 1000
                  ) + " km"
                : haversine(
                    this.props.currentLocation,
                    this.props.data.location[0]
                  ).toFixed(0) + " m")
            : "Undefined location"}
        </Text>
        <Text style={styles.description}>{this.props.data.description}</Text>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  title: {
    borderBottomWidth: 2,
    // borderBottomColor: '#e0e0e0',
    fontSize: 20,
    color: todoTitleColor,
    // padding: 5,
    fontWeight: "bold"
  }
});
