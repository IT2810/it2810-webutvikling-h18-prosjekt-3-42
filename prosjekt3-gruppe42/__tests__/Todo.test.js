import React from "react";
import {todos} from '../__mocks__/mock_todos';
import { haversine, toRadians } from "../functions";
import ToDoTitle from '../Components/ToDo/ToDoTitle';
import ToDoContent from '../Components/ToDo/ToDoContent';
import renderer from 'react-test-renderer';

describe("Test ToDoTitle component", ()=> {
    const tree = renderer.create(<ToDoTitle data={todos[0]} />);
    const tree2 = renderer.create(<ToDoTitle data={todos[4]} />);
    test("ToDoTitle snapshot test", () => {

        expect(tree).toMatchSnapshot();
    })

    test("Todo title should be in ToDoTitle", ()=> {
        expect(tree.toJSON().children[0]).toBe(todos[0].title);
    });

    test("Todo date should match date in ToDoTitle", ()=> {
        expect(tree.toJSON().children[3]).toEqual(expect.stringMatching(/[0-9]+.\s\w+\s\([0-9]{4}\)/))
    });

    test("Todo completed status should match in ToDoTitle", ()=> {
        expect(tree.toJSON().children[5]).toBe(" ");
        expect(tree2.toJSON().children[5]).not.toBe(" ");

    })
})

describe("Test ToDoContent component", ()=> {
    const tree = renderer.create(<ToDoContent data={todos[0]} currentLocation={{latitude:10, longitude:10}} />);
    test("Todo description should match description in ToDoContent ", ()=>{
        expect(tree.toJSON().children[1].children[0]).toBe(todos[0].description);
        // console.log(tree.toJSON().children[1].children[0])
    })

    test("ToDoContent distance should match distance between currentLocation and todo-location", ()=> {
        let p = /\d+/;
        let distance = p.exec(tree.toJSON().children[0].children[0])[0];
        expect(distance).toBe(String(Math.round((haversine(todos[0].location[0], {latitude:10, longitude:10}))/1000)))
    })
})
