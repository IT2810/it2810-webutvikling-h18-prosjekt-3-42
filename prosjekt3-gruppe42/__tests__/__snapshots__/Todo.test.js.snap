// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test ToDoTitle component ToDoTitle snapshot test 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Array [
      Object {
        "borderBottomWidth": 2,
        "color": "#424242",
        "fontSize": 20,
        "fontWeight": "bold",
      },
      Object {
        "borderBottomColor": "#CC2936",
      },
    ]
  }
>
  Sove
   - Deadline:
   
  16. oktober (2018)
   
   
</Text>
`;
